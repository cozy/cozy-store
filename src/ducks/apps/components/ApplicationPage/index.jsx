/* global cozy */
import React, { Component } from 'react'
import { Link } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import Button from 'cozy-ui/react/Button'
import FocusTrap from 'focus-trap-react'
import AppIcon from 'cozy-ui/react/AppIcon'

import ApplicationPageLoading from 'ducks/components/ApplicationPageLoading'
import { getTranslatedManifestProperty } from 'lib/helpers'
import Header from './Header'
import Gallery from './Gallery'
import Details from './Details'

import {
  preventBackgroundScroll,
  unpreventBackgroundScroll
} from 'lib/scrollHelpers.js'
import { getAppIconProps } from 'ducks/apps'

const MOBILE_PLATFORMS = ['ios', 'android']
const isMobilePlatform = name => MOBILE_PLATFORMS.includes(name)

const { BarCenter } = cozy.bar

export class ApplicationPage extends Component {
  constructor(props) {
    super(props)
    this.state = { displayBarIcon: false, activeTrap: true }
  }

  componentDidMount() {
    preventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.addEventListener('scroll', this.handleScroll, {
        passive: true
      })
    }
  }

  componentWillUnmount() {
    unpreventBackgroundScroll()
    const { mainPageRef } = this.props
    if (mainPageRef && mainPageRef.current) {
      mainPageRef.current.removeEventListener('scroll', this.handleScroll)
    }
  }

  mountTrap = () => {
    this.setState({ activeTrap: true })
  }

  unmountTrap = () => {
    this.setState({ activeTrap: false })
  }

  handleScroll = () => {
    if (
      !this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop > 100
    ) {
      this.setState(() => ({ displayBarIcon: true }))
    } else if (
      this.state.displayBarIcon &&
      this.props.mainPageRef.current.scrollTop < 100
    ) {
      this.setState(() => ({ displayBarIcon: false }))
    }
  }

  render() {
    const {
      t,
      parent,
      matchRoute,
      isFetching,
      fetchError,
      breakpoints = {},
      pauseFocusTrap,
      getApp,
      redirectTo
    } = this.props
    if (isFetching) return <ApplicationPageLoading />
    const app = getApp(matchRoute)
    if (!app) return redirectTo(`/${parent}`)

    const { displayBarIcon } = this.state
    const { isMobile } = breakpoints
    if (fetchError) {
      return (
        <p className="u-error">
          {t('app_modal.install.message.version_error', {
            message: fetchError.message
          })}
        </p>
      )
    }
    const { icon, slug, iconToLoad } = app
    const appName = getTranslatedManifestProperty(app, 'name', t)
    const namePrefix = getTranslatedManifestProperty(app, 'name_prefix', t)
    const appShortDesc = getTranslatedManifestProperty(
      app,
      'short_description',
      t
    )
    const appLongDesc = getTranslatedManifestProperty(
      app,
      'long_description',
      t
    )
    const appChanges = getTranslatedManifestProperty(app, 'changes', t)
    const mobileApps =
      app.platforms &&
      !!app.platforms.length &&
      app.platforms.reduce((mobilePlatforms, platformInfos) => {
        // force to be lower case
        platformInfos.type = platformInfos.type.toLowerCase()
        if (isMobilePlatform(platformInfos.type)) {
          mobilePlatforms.push(platformInfos)
        }
        return mobilePlatforms
      }, [])
    return (
      <FocusTrap
        focusTrapOptions={{
          onDeactivate: this.unmountTrap,
          clickOutsideDeactivates: true
        }}
        paused={pauseFocusTrap}
      >
        <div className="sto-modal-page-app">
          {isMobile && icon && !iconToLoad && (
            <BarCenter>
              <div className="sto-app-bar">
                <AppIcon
                  app={app}
                  className={`sto-app-bar-icon ${
                    !displayBarIcon ? 'sto-app-bar-icon--hidden' : ''
                  }`}
                  {...getAppIconProps()}
                />
              </div>
            </BarCenter>
          )}
          <div className="sto-app">
            <Button
              icon="back"
              tag={Link}
              to={`${parent}`}
              className="sto-app-back"
              label={t('app_page.back')}
              onClick={this.unmountTrap}
              subtle
            />
            <Header
              app={app}
              namePrefix={namePrefix}
              name={appName}
              description={appShortDesc}
              parent={parent}
            />
            {app.screenshots && !!app.screenshots.length && (
              <Gallery slug={slug} images={app.screenshots} />
            )}
            <Details
              app={app}
              description={appLongDesc}
              changes={appChanges}
              mobileApps={mobileApps}
              parent={parent}
            />
          </div>
        </div>
      </FocusTrap>
    )
  }
}

// translate is needed here for the lang props
export default translate()(withBreakpoints()(ApplicationPage))
