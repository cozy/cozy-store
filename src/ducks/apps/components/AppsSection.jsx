import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import { getTranslatedManifestProperty } from 'lib/helpers'

import SmallAppItem from 'ducks/components/SmallAppItem'

const _renderAppComponent = (app, t, onAppClick, isMobile) => {
  return (
    <SmallAppItem
      app={app}
      namePrefix={getTranslatedManifestProperty(app, 'name_prefix', t)}
      name={getTranslatedManifestProperty(app, 'name', t)}
      onClick={() => onAppClick(app.slug)}
      key={app.slug}
      isMobile={isMobile}
    />
  )
}

export const AppsSection = ({
  t,
  appsList,
  subtitle,
  onAppClick,
  breakpoints = {}
}) => {
  const { isMobile } = breakpoints
  return (
    <div className="sto-sections-apps">
      {subtitle}
      {appsList && !!appsList.length && (
        <div className="sto-sections-list">
          {appsList.map(app =>
            _renderAppComponent(app, t, onAppClick, isMobile)
          )}
        </div>
      )}
    </div>
  )
}

export default translate()(withBreakpoints()(AppsSection))
