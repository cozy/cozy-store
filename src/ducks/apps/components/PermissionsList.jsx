import React from 'react'

import { translate } from 'cozy-ui/react/I18n'
import Icon from 'cozy-ui/react/Icon'
import ReactMarkdownWrapper from '../../components/ReactMarkdownWrapper'

import localAccessIcon from 'assets/icons/icon-cloud-in-cozy.svg'
import externalIcon from 'assets/icons/icon-cloud-out-cozy.svg'

import LINXO_CONNECTORS from 'config/linxo.json'

export const Permission = ({ description, label, type, t }) => (
  <li key={type} className="sto-perm-list-item">
    <div className="sto-perm-label">
      <span className="sto-perm-title" data-doctype={type}>
        {label}
      </span>
      {description && (
        <small className="sto-perm-description">
          <ReactMarkdownWrapper
            className="sto-perm-description-detail"
            source={`${t('permissions.reason')}${description}`}
          />
        </small>
      )}
    </div>
  </li>
)

const LocalizedPermission = ({ t, slug, name, type }) => (
  <Permission
    description={t(`apps.${slug}.permissions.${name}.description`, { _: '' })}
    label={t(`doctypes.${type}`)}
    type={type}
    t={t}
  />
)

const getProcessedPermissions = (t, app) => {
  const external = [] // external permissions
  const externalTypes = [] // use to filter the internal after processing
  if (LINXO_CONNECTORS.includes(app.slug)) {
    const linxoType = 'io.cozy.accounts'
    external.push(
      <Permission
        description={t('permissions.linxo')}
        label={t(`doctypes.${linxoType}`)}
        type={linxoType}
        t={t}
      />
    )
    externalTypes.push(linxoType)
  }

  // internal permissions
  const internal = Object.keys(app.permissions || {})
    .filter(name => !externalTypes.includes(app.permissions[name].type))
    .map(name => (
      <LocalizedPermission
        key={name}
        {...app.permissions[name]}
        slug={app.slug}
        name={name}
        t={t}
      />
    ))
  return { internalPermissions: internal, externalPermissions: external }
}

export const PermissionsList = ({ t, app, appName }) => {
  const { externalPermissions, internalPermissions } = getProcessedPermissions(
    t,
    app
  )
  const developerName =
    (app.developer && app.developer.name) || t('permissions.developer_default')
  return app.permissions ? (
    <div>
      {!!externalPermissions.length && (
        <div className="sto-perm-list-title">
          <Icon
            className="sto-perm-list-icon"
            icon={externalIcon}
            size="24"
            color="#FF7F1B"
          />
          <p>
            {t('permissions.description.external', { name: developerName })}
          </p>
        </div>
      )}
      <ul className="sto-perm-list">{externalPermissions}</ul>
      {!!internalPermissions.length && (
        <div className="sto-perm-list-title">
          <Icon
            className="sto-perm-list-icon"
            icon={localAccessIcon}
            size="24"
            color="#35CE68"
          />
          <p>
            {t('permissions.description.internal', { name: developerName })}
          </p>
        </div>
      )}
      <ul className="sto-perm-list">{internalPermissions}</ul>
    </div>
  ) : (
    <ReactMarkdownWrapper
      source={t('permissions.description.nothing', {
        appName
      })}
    />
  )
}

export default translate()(PermissionsList)
