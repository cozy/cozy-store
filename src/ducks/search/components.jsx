import StoreAppItem from 'ducks/apps/components/StoreAppItem'
import { dumpMatches } from 'ducks/search/utils'
import sortBy from 'lodash/sortBy'
import PropTypes from 'prop-types'
import React, { useCallback, useMemo } from 'react'

import flag from 'cozy-flags'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import Icon from 'cozy-ui/transpiled/react/Icon'
import Magnifier from 'cozy-ui/transpiled/react/Icons/Magnifier'
import Input from 'cozy-ui/transpiled/react/Input'
import InputGroup from 'cozy-ui/transpiled/react/InputGroup'
import Label from 'cozy-ui/transpiled/react/Label'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

export const SearchField = ({ onChange, value }) => {
  const { t } = useI18n()
  const { isMobile } = useBreakpoints()
  const handleChange = useCallback(
    ev => {
      onChange(ev.target.value)
    },
    [onChange]
  )
  return (
    <div className="u-flex u-flex-items-center u-flex-grow-1">
      {!isMobile ? (
        <Label className="u-di u-mr-half" htmlFor="discover-search">
          {t('discover-search-field.label')}
        </Label>
      ) : null}
      <InputGroup
        prepend={
          isMobile ? (
            <Icon icon={Magnifier} className="u-pl-1 u-coolGrey" />
          ) : null
        }
      >
        <Input
          id="discover-search"
          placeholder={t('discover-search-field.placeholder')}
          onChange={handleChange}
          type="text"
          value={value}
        />
      </InputGroup>
    </div>
  )
}

export const SearchResults = ({ searchResults, onAppClick, disableClick }) => {
  const sortedSortResults = useMemo(() => {
    return sortBy(searchResults, result => result.score)
  }, [searchResults])
  return (
    <div className="u-mv-1 u-flex u-flex-wrap">
      {sortedSortResults.map(result => {
        const app = result.item
        const isDisableClick = disableClick?.(app)
        return flag('store.show-search-score') ? (
          <div>
            <StoreAppItem
              onClick={() => !isDisableClick && onAppClick(app.slug)}
              key={app.slug}
              app={app}
            />
            <small
              onClick={() => dumpMatches(result)}
              title={JSON.stringify(result.matches, null, 2)}
            >
              {result.score.toFixed(2)}
            </small>
          </div>
        ) : (
          <StoreAppItem
            onClick={() => !isDisableClick && onAppClick(app.slug)}
            key={app.slug}
            app={app}
          />
        )
      })}
    </div>
  )
}

SearchResults.propTypes = {
  onAppClick: PropTypes.func.isRequired,
  disableClick: PropTypes.func,
  searchResults: PropTypes.array
}
