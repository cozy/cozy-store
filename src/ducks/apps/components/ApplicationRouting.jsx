import React, { Component } from 'react'
import { Route, withRouter } from 'react-router-dom'

import IntentModal from 'cozy-ui/react/IntentModal'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'

import PermissionsModal from './PermissionsModal'
import UninstallModal from './UninstallModal'
import Install from './Install'
import ApplicationPage from './ApplicationPage'

import { APP_TYPE, REGISTRY_CHANNELS } from 'ducks/apps'

export class ApplicationRouting extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isInstallSuccess: false
    }
    this.getAppFromMatchOrSlug = this.getAppFromMatchOrSlug.bind(this)
  }

  componentWillReceiveProps(nextProps) {
    // don't react on error
    if (nextProps.actionError) return
    // on install success
    if (this.props.isInstalling && !nextProps.isInstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      if (app.type === APP_TYPE.KONNECTOR) {
        return history.replace(`/${parent}/${app.slug}/configure`)
      } else {
        Alerter.success(t('app_modal.install.message.install_success'), {
          duration: 3000
        })
        return history.replace(`/${parent}/${app.slug}`)
      }
    } else if (this.props.isUninstalling && !nextProps.isUninstalling) {
      const { history, location, parent, t } = this.props
      const pathRegex = new RegExp(`^/${parent}/([^/]*)/.*`)
      const matches = location.pathname.match(pathRegex)
      if (!matches || matches.length < 1) return history.replace(`/${parent}/`)
      const app = this.getAppFromMatchOrSlug(null, matches[1])
      Alerter.success(t('app_modal.uninstall.message.success'), {
        duration: 3000
      })
      return history.replace(`/${parent}/${app.slug}`)
    }
  }

  getAppFromMatchOrSlug(match, slug) {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (match && match.params && match.params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  render() {
    const {
      fetchLatestApp,
      installApp,
      uninstallApp,
      updateApp,
      isFetching,
      isAppFetching,
      isInstalling,
      parent,
      history,
      actionError,
      fetchError
    } = this.props
    return (
      <div>
        <Route
          path={`/${parent}/:appSlug`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <ApplicationPage app={app} parent={parent} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/channel/:channel`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app || !app.isInRegistry) return history.replace(`/${parent}`)
            const channel = match.params.channel
            const isChannelAvailable = Object.values(
              REGISTRY_CHANNELS
            ).includes(channel)
            if (!isChannelAvailable) {
              return history.replace(`/${parent}/${app.slug}/manage`)
            }
            return (
              <Install
                installApp={app.installed ? updateApp : installApp}
                parent={`/${parent}`}
                fetchApp={chan => fetchLatestApp(app.slug, chan)}
                isAppFetching={isAppFetching}
                installError={actionError}
                fetchError={fetchError}
                app={app}
                isInstalling={isInstalling}
                channel={channel}
              />
            )
          }}
        />
        <Route
          path={`/${parent}/:appSlug/manage`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            if (app && app.installed) {
              if (!app.uninstallable)
                return history.replace(`/${parent}/${app.slug}`)
              return (
                <UninstallModal
                  uninstallApp={uninstallApp}
                  parent={`/${parent}`}
                  uninstallError={actionError}
                  app={app}
                />
              )
            } else {
              return (
                <Install
                  installApp={installApp}
                  parent={`/${parent}`}
                  installError={actionError}
                  app={app}
                  isInstalling={isInstalling}
                />
              )
            }
          }}
        />
        <Route
          path={`/${parent}/:appSlug/permissions`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            return <PermissionsModal app={app} parent={`/${parent}`} />
          }}
        />
        <Route
          path={`/${parent}/:appSlug/configure`}
          render={({ match }) => {
            if (isFetching) return
            const app = this.getAppFromMatchOrSlug(match)
            if (!app) return history.replace(`/${parent}`)
            const goToApp = () => history.replace(`/${parent}/${app.slug}`)
            if (app && app.installed && app.type === APP_TYPE.KONNECTOR) {
              return (
                <IntentModal
                  action="CREATE"
                  doctype="io.cozy.accounts"
                  options={{ slug: app.slug }}
                  dismissAction={goToApp}
                  onComplete={goToApp}
                  mobileFullscreen
                  overflowHidden
                  size="small"
                  height="35rem"
                />
              )
            } else {
              return goToApp()
            }
          }}
        />
      </div>
    )
  }
}

export default withRouter(translate()(ApplicationRouting))
