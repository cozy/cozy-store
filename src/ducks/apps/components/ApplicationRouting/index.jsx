import ApplicationPage from 'ducks/apps/components/ApplicationPage'
import ChannelRoute from 'ducks/apps/components/ApplicationRouting/ChannelRoute'
import ConfigureRoute from 'ducks/apps/components/ApplicationRouting/ConfigureRoute'
import InstallRoute from 'ducks/apps/components/ApplicationRouting/InstallRoute'
import PermissionsRoute from 'ducks/apps/components/ApplicationRouting/PermissionsRoute'
import UninstallRoute from 'ducks/apps/components/ApplicationRouting/UninstallRoute'
import React, { Component } from 'react'
import {
  Outlet,
  Route,
  Routes,
  useLocation,
  useNavigate
} from 'react-router-dom'

import { translate } from 'cozy-ui/transpiled/react/I18n'

const OutletWrapper = ({ Component }) => (
  <>
    <Component />
    <Outlet />
  </>
)

export class ApplicationRouting extends Component {
  mainPage = React.createRef()

  getAppFromMatchOrSlug = (params, slug) => {
    const appsArray = this.props.apps || this.props.installedApps || []
    const appSlug = slug || (params && params.appSlug)
    if (!appsArray.length || !appSlug) return null
    const app = appsArray.find(app => app.slug === appSlug)
    return app
  }

  redirectTo = target => {
    const { navigate, location } = this.props
    navigate(target + location.search, { replace: true })
    return null
  }

  render() {
    const { isFetching, parent } = this.props
    return (
      <div className="sto-modal-page" ref={this.mainPage}>
        <Routes>
          <Route
            path=":appSlug"
            element={
              <OutletWrapper
                Component={() => (
                  <ApplicationPage
                    parent={parent}
                    isFetching={isFetching}
                    getApp={this.getAppFromMatchOrSlug}
                    redirectTo={this.redirectTo}
                    mainPageRef={this.mainPage}
                  />
                )}
              />
            }
          >
            <Route
              path="channel/:channel"
              element={
                <ChannelRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="install"
              element={
                <InstallRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="uninstall"
              element={
                <UninstallRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="permissions"
              element={
                <PermissionsRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
            <Route
              path="configure"
              element={
                <ConfigureRoute
                  getApp={this.getAppFromMatchOrSlug}
                  isFetching={isFetching}
                  parent={parent}
                  redirectTo={this.redirectTo}
                />
              }
            />
          </Route>
        </Routes>
      </div>
    )
  }
}

const ApplicationRoutingWrapper = props => {
  const navigate = useNavigate()
  const location = useLocation()
  return (
    <ApplicationRouting {...props} navigate={navigate} location={location} />
  )
}

export default translate()(ApplicationRoutingWrapper)
