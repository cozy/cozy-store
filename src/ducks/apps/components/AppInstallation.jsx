import React, { Component } from 'react'

import { connect } from 'react-redux'
import { PropTypes } from 'react-proptypes'
import flags from 'cozy-flags'

import ReactMarkdownWrapper from 'ducks/components/ReactMarkdownWrapper'
import { ModalDescription, ModalHeader, ModalFooter } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'
import Button from 'cozy-ui/react/Button'

import PermissionsList from 'ducks/apps/components/PermissionsList'
import Partnership from 'ducks/apps/components/Partnership'
import { translate } from 'cozy-ui/react/I18n'
import Alerter from 'cozy-ui/react/Alerter'
import Checkbox from 'cozy-ui/react/Checkbox'

import { getTranslatedManifestProperty } from 'lib/helpers'
import { hasPendingUpdate } from 'ducks/apps/appStatus'
import storeConfig from 'config'

import { APP_TYPE, getAppBySlug, installAppFromRegistry } from 'ducks/apps'

const shouldSkipPermissions = app =>
  flags('skip-low-permissions') &&
  app.label <= storeConfig.default.authorizedLabelLimit

const hasInstallation = app => !app.installed || hasPendingUpdate(app)

export class AppInstallation extends Component {
  constructor(props) {
    super(props)
    const { app } = props
    this.state = { isTermsAccepted: false }
    this.installApp = this.installApp.bind(this)
    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      this.installApp()
    }
  }

  installApp() {
    if (!this.isInstallReady()) return // Not ready to be installed
    const { app, channel, installApp, onInstallOrUpdate, t } = this.props
    const isUpdate = app.installed
    const isChannelSwitch = !!channel
    installApp(app, channel, isUpdate).catch(error => {
      const appName = getTranslatedManifestProperty(app, 'name', t)
      Alerter.error(
        t('app_modal.install.alert.install_error', {
          message: error.message,
          name: appName
        }),
        {
          duration: 12000,
          buttonText: t('app_modal.install.alert.dismiss'),
          buttonAction: dismiss => dismiss()
        }
      )
      console.error(error)
    })
    if (isUpdate || isChannelSwitch || app.type !== APP_TYPE.KONNECTOR) {
      if (typeof onInstallOrUpdate === 'function') onInstallOrUpdate()
    }
  }

  componentDidUpdate = prevProps => {
    const { app, onKonnectorInstall } = this.props
    const justInstalled =
      !!prevProps.app && !prevProps.app.installed && app.installed
    if (justInstalled && app.type === APP_TYPE.KONNECTOR) {
      if (typeof onKonnectorInstall === 'function') onKonnectorInstall()
    }
  }

  acceptTerms = () => {
    this.setState(state => ({
      isTermsAccepted: !state.isTermsAccepted
    }))
  }

  isInstallReady = () => {
    const { app, isInstalling } = this.props
    const isCurrentAppInstalling = isInstalling === app.slug
    const isTermsReady =
      shouldSkipPermissions(app) || (!app.terms || this.state.isTermsAccepted)
    return !isCurrentAppInstalling && isTermsReady
  }

  render() {
    const {
      app,
      fetchError,
      installError,
      isFetching,
      isAppFetching,
      isInstalling,
      onCancel,
      t
    } = this.props
    const { isTermsAccepted } = this.state
    const isFetchingSomething = isFetching || isAppFetching
    const isCurrentAppInstalling = isInstalling === app.slug

    if (shouldSkipPermissions(app) && hasInstallation(app)) {
      return (
        <React.Fragment>
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xxlarge" />
            </div>
          </ModalDescription>
        </React.Fragment>
      )
    }

    return (
      <React.Fragment>
        <ModalHeader title={t('app_modal.install.title')} />
        {isFetchingSomething ? (
          <ModalDescription className="sto-install-loading-wrapper">
            <div className="sto-install-loading">
              <Spinner size="xlarge" />
            </div>
          </ModalDescription>
        ) : (
          <ModalDescription>
            {app.partnership && <Partnership app={app} />}
            {app.permissions && <PermissionsList app={app} />}
            {fetchError && (
              <p className="u-error">
                {t('app_modal.install.message.version_error', {
                  message: fetchError.message
                })}
              </p>
            )}
          </ModalDescription>
        )}
        {!isFetchingSomething && !fetchError && (
          <ModalFooter className="sto-install-footer">
            {installError && (
              <p className="u-error">
                {t('app_modal.install.message.install_error', {
                  message: installError.message
                })}
              </p>
            )}
            {app.terms && (
              <div className="sto-install-terms">
                <Checkbox
                  className="sto-install-terms-checkbox"
                  onChange={this.acceptTerms}
                  checked={isTermsAccepted}
                  disabled={isInstalling}
                >
                  <ReactMarkdownWrapper
                    source={t('app_modal.install.terms', {
                      url: app.terms.url
                    })}
                  />
                </Checkbox>
              </div>
            )}
            <div className="sto-install-controls">
              <Button
                theme="secondary"
                onClick={onCancel}
                disabled={isCurrentAppInstalling}
                label={t('app_modal.install.cancel')}
                extension="full"
                className="u-mh-half"
              />
              <Button
                theme="primary"
                disabled={!this.isInstallReady() || isInstalling}
                busy={isCurrentAppInstalling}
                extension="full"
                onClick={this.installApp}
                label={
                  hasPendingUpdate(app)
                    ? t('app_modal.install.update')
                    : t('app_modal.install.install')
                }
                className="u-mh-half"
              />
            </div>
          </ModalFooter>
        )}
      </React.Fragment>
    )
  }
}

AppInstallation.propTypes = {
  appSlug: PropTypes.string.isRequired
}

const mapStateToProps = (state, ownProps) => ({
  app: getAppBySlug(state, ownProps.appSlug),
  isFetching: state.apps.isFetching,
  isAppFetching: state.apps.isAppFetching,
  isInstalling: state.apps.isInstalling,
  fetchError: state.apps.fetchError,
  installError: state.apps.actionError
})

const mapDispatchToProps = dispatch => ({
  installApp: (app, channel, isUpdate) =>
    dispatch(installAppFromRegistry(app, channel, isUpdate))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AppInstallation))
