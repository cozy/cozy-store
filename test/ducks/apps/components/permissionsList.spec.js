'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { tMock } from '../../../jestLib/I18n'
import { PermissionsList } from 'ducks/apps/components/PermissionsList'

import mockAppVersion from '../_mockPhotosRegistryVersion'
import mockKonnectorVersion from '../_mockPKonnectorTrinlaneRegistryVersion'
import mockBankKonnectorWithExternalDoctype from '../_mockBankKonnectorWithExternalDoctype'

describe('PermissionsList component', () => {
  beforeAll(() => {
    // define global mock url
    global.cozy = {
      client: {
        _url: '//cozytest.mock.cc'
      }
    }
  })

  it('should be rendered correctly with permissions', () => {
    const component = shallow(
      <PermissionsList t={tMock} app={mockAppVersion.manifest} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly with konnector permissions', () => {
    const component = shallow(
      <PermissionsList t={tMock} app={mockKonnectorVersion.manifest} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly without permissions', () => {
    const component = shallow(
      <PermissionsList t={tMock} app={{ name: 'Mock' }} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly when is a bank with external doctypes permissions', () => {
    const component = shallow(
      <PermissionsList t={tMock} app={mockBankKonnectorWithExternalDoctype} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
