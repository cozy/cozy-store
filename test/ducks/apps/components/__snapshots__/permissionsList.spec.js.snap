// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionsList component should be rendered correctly with konnector permissions 1`] = `
<div>
  <ul
    className="sto-perm-list"
  />
  <div
    className="sto-perm-list-title"
  >
    <Icon
      className="sto-perm-list-icon"
      color="#35CE68"
      icon="test-file-stub"
      size="24"
      spin={false}
    />
    <p>
      Data can be accessed but neither shared nor transmitted outside your Cozy:
    </p>
  </div>
  <ul
    className="sto-perm-list"
  >
    <LocalizedPermission
      description="Required to link bills to bank operations"
      name="bank operations"
      slug="konnector-trinlane"
      t={[Function]}
      type="io.cozy.bank.operations"
    />
    <LocalizedPermission
      description="Required to save the bills data"
      name="bills"
      slug="konnector-trinlane"
      t={[Function]}
      type="io.cozy.bills"
    />
    <LocalizedPermission
      description="Required to save the bills"
      name="files"
      slug="konnector-trinlane"
      t={[Function]}
      type="io.cozy.files"
    />
    <LocalizedPermission
      description="Required to get the account's data"
      name="accounts"
      slug="konnector-trinlane"
      t={[Function]}
      type="io.cozy.accounts"
      verbs={
        Array [
          "GET",
        ]
      }
    />
  </ul>
</div>
`;

exports[`PermissionsList component should be rendered correctly with permissions 1`] = `
<div>
  <ul
    className="sto-perm-list"
  />
  <div
    className="sto-perm-list-title"
  >
    <Icon
      className="sto-perm-list-icon"
      color="#35CE68"
      icon="test-file-stub"
      size="24"
      spin={false}
    />
    <p>
      Data can be accessed but neither shared nor transmitted outside your Cozy:
    </p>
  </div>
  <ul
    className="sto-perm-list"
  >
    <LocalizedPermission
      description="Required for photo access"
      methods={
        Array [
          "GET",
          "POST",
          "PUT",
        ]
      }
      name="files"
      slug="photos"
      t={[Function]}
      type="io.cozy.files"
    />
    <LocalizedPermission
      description="Required by the cozy-bar to display the icons of the apps"
      name="apps"
      slug="photos"
      t={[Function]}
      type="io.cozy.apps"
      verbs={
        Array [
          "GET",
        ]
      }
    />
    <LocalizedPermission
      description="Required to manage photos albums"
      methods={
        Array [
          "GET",
          "POST",
          "PUT",
        ]
      }
      name="albums"
      slug="photos"
      t={[Function]}
      type="io.cozy.photos.albums"
    />
    <LocalizedPermission
      description="Required to to share photos with your contacts"
      methods={
        Array [
          "GET",
          "POST",
        ]
      }
      name="contacts"
      slug="photos"
      t={[Function]}
      type="io.cozy.contacts"
    />
    <LocalizedPermission
      description="Required by the cozy-bar to display Claudy and know which applications are coming soon"
      name="settings"
      slug="photos"
      t={[Function]}
      type="io.cozy.settings"
      verbs={
        Array [
          "GET",
        ]
      }
    />
  </ul>
</div>
`;

exports[`PermissionsList component should be rendered correctly without permissions 1`] = `
<ReactMarkdownWrapper
  source="%{appName} doesn't require any data access to work correctly on your Cozy."
/>
`;
